pkg/reconciler/v1alpha1/broker/broker.go

```diff
@@ -0,0 +1,442 @@
+}
+
+// updateStatus may in fact update the broker's finalizers in addition to the status.
+func (r *reconciler) updateStatus(broker *v1alpha1.Broker) (*v1alpha1.Broker, error) {
+	ctx := context.TODO()
+	objectKey := client.ObjectKey{Namespace: broker.Namespace, Name: broker.Name}
+	latestBroker := &v1alpha1.Broker{}
+
+	if err := r.client.Get(ctx, objectKey, latestBroker); err != nil {
+		return nil, err
+	}
+
+	brokerChanged := false
+
+	if !equality.Semantic.DeepEqual(latestBroker.Finalizers, broker.Finalizers) {
+		latestBroker.SetFinalizers(broker.ObjectMeta.Finalizers)
+		if err := r.client.Update(ctx, latestBroker); err != nil {
+			return nil, err
+		}
+		brokerChanged = true
+	}
+
+	if equality.Semantic.DeepEqual(latestBroker.Status, broker.Status) {
+		return latestBroker, nil
+	}
+
+	if brokerChanged {
+		// Re-fetch.
+		latestBroker = &v1alpha1.Broker{}
+		if err := r.client.Get(ctx, objectKey, latestBroker); err != nil {
+			return nil, err
+		}
+	}
+
+	latestBroker.Status = broker.Status
+	if err := r.client.Status().Update(ctx, latestBroker); err != nil {
+		return nil, err
+	}
+
+	return latestBroker, nil
+}
+
+// reconcileFilterDeployment reconciles Broker's 'b' filter deployment.
+func (r *reconciler) reconcileFilterDeployment(ctx context.Context, b *v1alpha1.Broker) (*v1.Deployment, error) {
+	expected := resources.MakeFilterDeployment(&resources.FilterArgs{
+		Broker:             b,
+		Image:              r.filterImage,
+		ServiceAccountName: r.filterServiceAccountName,
+	})
+	return r.reconcileDeployment(ctx, expected)
+}
+
+// reconcileFilterService reconciles Broker's 'b' filter service.
+func (r *reconciler) reconcileFilterService(ctx context.Context, b *v1alpha1.Broker) (*corev1.Service, error) {
+	expected := resources.MakeFilterService(b)
+	return r.reconcileService(ctx, expected)
+}
+
+// reconcileChannel reconciles Broker's 'b' underlying channel.
+func (r *reconciler) reconcileChannel(ctx context.Context, b *v1alpha1.Broker) (*v1alpha1.Channel, error) {
+	c, err := r.getChannel(ctx, b)
+	// If the resource doesn't exist, we'll create it
+	if k8serrors.IsNotFound(err) {
+		c = newChannel(b)
+		err = r.client.Create(ctx, c)
+		if err != nil {
+			return nil, err
+		}
+		return c, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	// TODO Determine if we want to update spec (maybe just args?).
+	// Update Channel if it has changed. Note that we need to both ignore the real Channel's
+	// subscribable section and if we need to update the real Channel, retain it.
+	//expected.Spec.Subscribable = c.Spec.Subscribable
+	//if !equality.Semantic.DeepDerivative(expected.Spec, c.Spec) {
+	//	c.Spec = expected.Spec
+	//	err = r.client.Update(ctx, c)
+	//	if err != nil {
+	//		return nil, err
+	//	}
+	//}
+	return c, nil
+}
+
+// getChannel returns the Channel object for Broker 'b' if exists, otherwise it returns an error.
+func (r *reconciler) getChannel(ctx context.Context, b *v1alpha1.Broker) (*v1alpha1.Channel, error) {
```

Comment:
There are some edge cases here due to the way Get is implemented.
Current selection is based on LabelSelector. If we have more than one channel with the same labels then with every reconcile a different channel could have been read and Broker status would be either NotReady or Ready depending on which channel was at first index. However the ingress service continues to use the channel that was read when reconcile ran for first time. This could make debugging hard.

Suggestion:  Since we don't expect users who use Broker, to create their own channels and then configure it with the broker, why don't we Get one single channel based on its name because the Broker created the Channel.

Thoughts?